File: IntersectionOfTwoLists.cpp
// Time Complexity : o(n)
// Space Complexity : o(1)
// Did this code successfully run on Leetcode : YES
// Any problem you faced while coding this : NO

class Solution {
public:
    int getLen(ListNode *x) {
        int len = 0;
        while(x != NULL) {
            x=x->next;
            ++len;
        }
        return len;
    }
    ListNode *moveByDist(ListNode *x, int dist) {
        while(dist--) {
            x=x->next;
        }
        return x;
    }
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        int aLen = getLen(headA);
        int bLen = getLen(headB);
        if(aLen > bLen) {
            headA = moveByDist(headA, aLen-bLen);
        } else if(bLen > aLen) {
            headB = moveByDist(headB, bLen-aLen);
        }
        
        while(headA!=NULL && headB!=NULL) {
            if(headA==headB) {
                return headA;
            }
            headA=headA->next;
            headB=headB->next;
        }
        return NULL;
    }
};

File: intersection of two lists.py
# // Time Complexity : O(N)
# // Space Complexity : O(1)
# // Did this code successfully run on Leetcode : YES
# // Any problem you faced while coding this : No


# // Your code here along with comments explaining your approach
class Solution:
    def skipHead(self, head, length):
        while length:
            head = head.next
            length -= 1
        return head

    def getIntersectionNode(self, headA, headB):
        len_A, len_B = 0, 0

        temp = headA
        while temp:
            temp = temp.next
            len_A += 1

        temp = headB
        while temp:
            temp = temp.next
            len_B += 1

        if len_A > len_B:
            headA = self.skipHead(headA, len_A - len_B)
        elif len_B > len_A:
            headB = self.skipHead(headB, len_B - len_A)

        while headA and headB:
            if headA == headB:
                return headA
            headA = headA.next
            headB = headB.next

        return None
